feat add ipmi component (#4808)
grafana/alloy: docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md

At least one `target` block or the `local` block must be configured.

[target]: #target
[local]: #local

### `target`

| Name        | Type     | Description                                | Default     | Required |
| ----------- | -------- | ------------------------------------------ | ----------- | -------- |
| `address`   | `string` | IP address or hostname of the IPMI device. |             | yes      |
| `name`      | `string` | Name of the target used in job label.      |             | yes      |
| `driver`    | `string` | IPMI driver to use, `LAN_2_0` or `LAN`.    | `"LAN_2_0"` | no       |
| `module`    | `string` | IPMI collector module to use.              |             | no       |
| `password`  | `secret` | IPMI password for authentication.          |             | no       |
| `privilege` | `string` | IPMI privilege level, `user` or `admin`.   | `"admin"`   | no       |
| `user`      | `string` | IPMI username for authentication.          |             | no       |

The `target` block defines a remote IPMI device to monitor.
You can specify multiple `target` blocks to monitor multiple devices.

The `driver` attribute specifies which IPMI protocol version to use:

> - `LAN_2_0`: IPMI 2.0 protocol, recommended for most modern hardware

Author: I followed the upstream ipmi_exporter convention, LAN and LAN_2_0 are the driver names passed directly to freeipmi. I think we should keep it as it is.

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.ipmi/
aliases:
  - ../prometheus.exporter.ipmi/ # /docs/alloy/latest/reference/components/prometheus.exporter.ipmi/

> description: Learn about prometheus.exporter.ipmi

Author: Suggestion applied

labels:
  stage: general-availability
  products:
    - oss

> title: prometheus.exporter.ipmi

Author: Suggestion applied

## Usage

```alloy
prometheus.exporter.ipmi "<LABEL>" {
  target {
    name    = "<NAME>"
    address = "<IPMI_ADDRESS>"
    user    = "<USERNAME>"
    password = "<PASSWORD>"
  }
}
```

## Arguments

You can use the following arguments with `prometheus.exporter.ipmi`:

| Name          | Type       | Description                                           | Default | Required |
| ------------- | ---------- | ----------------------------------------------------- | ------- | -------- |
| `timeout`     | `duration` | Timeout for IPMI requests.                            | `"30s"` | no       |
| `config_file` | `string`   | Path to IPMI exporter configuration file.            |         | no       |
| `ipmi_config` | `string`   | IPMI exporter configuration as inline YAML string.    |         | no       |

The `timeout` should be set high enough to allow IPMI sensor collection to complete.
IPMI operations can be slow, especially for devices with many sensors.
A timeout of 30 seconds or more is recommended for most hardware.

The `config_file` and `ipmi_config` arguments are mutually exclusive.

> They allow advanced configuration of IPMI collection modules and command overrides.

Reviewer: ```suggestion
These arguments enable advanced configuration of IPMI collection modules and command overrides.
```

Author: Suggestion applied

| [`local`][local]   | Configures local IPMI collection.    | no       |

At least one `target` block or the `local` block must be configured.

[target]: #target
[local]: #local

### `target`

| Name        | Type     | Description                                   | Default     | Required |
| ----------- | -------- | --------------------------------------------- | ----------- | -------- |
| `name`      | `string` | Name of the target (used in job label).      |             | yes      |
| `address`   | `string` | IP address or hostname of the IPMI device.    |             | yes      |
| `user`      | `string` | IPMI username for authentication.             |             | no       |
| `password`  | `secret` | IPMI password for authentication.             |             | no       |
| `driver`    | `string` | IPMI driver to use (`LAN_2_0` or `LAN`).      | `"LAN_2_0"` | no       |
| `privilege` | `string` | IPMI privilege level (`user` or `admin`).    | `"admin"`   | no       |
| `module`    | `string` | IPMI collector module to use.                 |             | no       |

The `target` block defines a remote IPMI device to monitor.

> Multiple `target` blocks can be specified to monitor multiple devices.

Reviewer: ```suggestion
You can specify multiple `target` blocks to monitor multiple devices.
```

The `driver` attribute specifies which IPMI protocol version to use:
- `LAN_2_0`: IPMI 2.0 protocol (recommended for most modern hardware)
- `LAN`: IPMI 1.5 protocol (for older hardware)

The `privilege` attribute sets the privilege level for IPMI operations:
- `admin`: Full administrative access (required for all sensor data on most hardware)
- `user`: User-level access (limited sensor access)

### `local`

| Name      | Type     | Description                              | Default | Required |
| --------- | -------- | ---------------------------------------- | ------- | -------- |
| `enabled` | `bool`   | Enable local IPMI collection.            | `false` | no       |
| `module`  | `string` | IPMI collector module to use.            |         | no       |

The `local` block enables collection from the local machine's IPMI interface.
This requires the machine to have IPMI hardware and appropriate permissions.
On Linux systems, this typically requires:
- FreeIPMI tools installed
- Kernel modules loaded (`ipmi_devintf`, `ipmi_si`)

> - Appropriate device permissions (`/dev/ipmi0`)

Reviewer: ```suggestion
On Linux systems, this typically requires:

- FreeIPMI tools installed
- Kernel modules for `ipmi_devintf` and `ipmi_si` loaded
- Appropriate device permissions for `/dev/ipmi0`
```

## Collected metrics

The IPMI exporter collects the following metrics:

| Metric                      | Type  | Description                                                       |
| --------------------------- | ----- | ----------------------------------------------------------------- |
| `ipmi_up`                   | Gauge | IPMI device reachability (1=up, 0=down)                           |
| `ipmi_temperature_celsius`  | Gauge | Temperature sensor reading in degrees Celsius                     |
| `ipmi_fan_speed_rpm`        | Gauge | Fan speed sensor reading in RPM                                   |
| `ipmi_voltage_volts`        | Gauge | Voltage sensor reading in volts                                   |
| `ipmi_power_watts`          | Gauge | Power sensor reading in watts                                     |
| `ipmi_current_amperes`      | Gauge | Current sensor reading in amperes                                 |
| `ipmi_sensor_state`         | Gauge | Sensor state (0=nominal, 1=warning, 2=critical, 3=not available)  |
| `ipmi_collector_info`       | Gauge | Information about the IPMI collector configuration               |

All sensor metrics include labels:

> - `target`: IP address or "localhost" for local collection

Reviewer: ```suggestion
All sensor metrics include labels:

- `target`: IP address or "localhost" for local collection
```

    address  = "192.168.1.100"
    user     = "monitoring"
    password = env("IPMI_PASSWORD")
  }
}
```

Metrics are available at:
```
http://localhost:12345/api/v0/component/prometheus.exporter.ipmi.test/metrics
```

## Technical notes

### IPMI timeout configuration

IPMI sensor collection can be slow, especially for servers with many sensors.
Configure timeouts appropriately:

- **Component timeout**: Set to 30-60 seconds to allow sensor collection to complete
- **Scrape timeout**: Must be longer than the component timeout
- **Scrape interval**: Recommended 60 seconds or more to avoid overlapping collections

Example configuration:
```alloy
prometheus.exporter.ipmi "servers" {
  timeout = "30s"  // Allow 30 seconds for IPMI operations
  // ... targets ...
}

prometheus.scrape "ipmi" {
  scrape_interval = "60s"  // Scrape every minute
  scrape_timeout  = "45s"  // Allow 45 seconds (> component timeout)
  targets         = prometheus.exporter.ipmi.servers.targets
  // ... forward_to ...
}
```

### IPMI driver selection

Choose the appropriate IPMI driver based on your hardware:

- **LAN_2_0** (IPMI 2.0): Recommended for modern servers (Dell, HP, Supermicro from ~2010+)

> - **LAN** (IPMI 1.5): For older hardware that doesn't support IPMI 2.0

Reviewer: ```suggestion
- **LAN** (IPMI 1.5): For legacy hardware that doesn't support IPMI 2.0
```

If unsure, try `LAN_2_0` first. Connection failures may indicate the need to use `LAN`.

### Local IPMI requirements

> For local IPMI collection, ensure:

Reviewer: ```suggestion
For local IPMI collection, ensure the following minimum requirements are met:
```

Author: Done

1. **Hardware support**: Server has BMC/IPMI hardware
2. **Kernel modules** (Linux):
   ```bash
   modprobe ipmi_devintf
   modprobe ipmi_si
   ```
3. **Device permissions**:
   ```bash
   ls -l /dev/ipmi* /dev/ipmi0
   # Should be accessible by the user running Alloy
   ```
4. **FreeIPMI tools** (optional but recommended):
   ```bash
   # Debian/Ubuntu
   apt-get install freeipmi-tools

   # RHEL/CentOS
   yum install freeipmi
   ```

### Security considerations

IPMI credentials provide low-level hardware access:

- **Use dedicated monitoring users** with minimal required privileges
- **Store passwords securely** using `env()` or secrets management
- **Never commit passwords** to version control

> - **Consider network isolation** for IPMI interfaces (dedicated management network)

Reviewer: ```suggestion
- **Consider network isolation** for IPMI interfaces on dedicated management networks
```

Author: Done
